sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ClientRetryPolicy.java

```diff
@@ -110,17 +110,17 @@ public Mono<ShouldRetryResult> shouldRetry(Exception e) {
                 this.isReadRequest)
         {
             logger.warn("Endpoint not available for reads. Will refresh cache and retry. ", e);
-            return this.shouldRetryOnEndpointFailureAsync(true, false);
+            return this.shouldRetryOnEndpointFailureAsync(true, false, false);
         }
 
         // Received Connection error (HttpRequestException), initiate the endpoint rediscovery
         if (WebExceptionUtility.isNetworkFailure(e)) {
             if (clientException != null && Exceptions.isSubStatusCode(clientException, HttpConstants.SubStatusCodes.GATEWAY_ENDPOINT_UNAVAILABLE)) {
                 if (this.isReadRequest || WebExceptionUtility.isWebExceptionRetriable(e)) {
                     logger.warn("Gateway endpoint not reachable. Will refresh cache and retry. ", e);
-                    return this.shouldRetryOnEndpointFailureAsync(this.isReadRequest, false);
+                    return this.shouldRetryOnEndpointFailureAsync(this.isReadRequest, false, isReadRequest);
```

Comment:
why only for `isReadRequest` -- I think we can safe pass `true` directly

Reply:
What if it is a write request and a retriable exception? I think writes failover flip-flops between 1st and 2nd write-enabled regions in case of multi-write account.

Reply:
I think for metadata request -> we are going to flip between the 1st and 2nd write region
and for others -> we are going to loop through all other write regions

Reply:
@jeet1995  we can create a github issue for tracking the change. Another change would be to mark both readEndpoint and writeEndpoint unavailable - to be consistent with .Net behavior